### DISCUSSION TITLE
How to integrate with boost::asio?

### DISCUSSION THREAD

**Discussion by SeeRich** [danvratil/qcoro#108] [General]
I currently have all my applications networking code written with boost::asio and returning futures. For example:
```c++
std::future<std::string> restRequest(const HttpRequest& req);
```
However, this doesn't integrate the best with my Qt UI and I would love to integrate this with QCoro.

If I modify the method above to return a boost coroutine instead:
```c++
boost::asio::awaitable<std::string> restRequest(const HttpRequest& req);
```
What would I need to do to `co_await` it inside a QCoro::Task<>?
e.g. (barrowed from QCoro networking example)
```c++
QCoro::Task<> start()
    {
        mPb->setVisible(true);
        mBtn->setEnabled(false);
        mBtn->setText(tr("Downloading ..."));

        auto request = HttpRequest();
        ... fill in request ....

        auto response = co_await restRequest(request);
        std::cout << response << std::endl;

        mPb->setVisible(false);
        mBtn->setEnabled(true);
        mBtn->setText(tr("Done, download again"));
    }
```

I did try this and the code doesn't compile because the `boost::asio::awaitable` type doesn't meet the requirements of `QCoro::Task`.

Also, I am somewhat concerned that this will not work anyways due to the `restRequest` using a `boost::asio::io_context` under the hood. This means the coroutine being awaited above will actually have its IO handled by a thread assigned to the `io_context`.

Any suggestions appreciated!

**Comment by danvratil** [comment]
In theory the `Promise::await_transform()` overload should accept a type that satisfies the QCoro's Awaitable concept. But it's fairly possible that the concept is incorrect or incomplete so the boost's awaitable is not accepted, I'll look into that.

What's more problematic is the integration of the two event loops - boost's `io_context` and the Qt event loop. This is a problem unrelated to QCoro - you would have the same problem even without coroutines. I think the only solution is to run the boost's `io_context` in a separate thread - which I believe is what you did previously with the `std::future`-based implementation?

If you are using Qt6, you could replace `std::future` with `QFuture` and `QPromise`. `QFuture` is supported by QCoro, so this would work out-of-the-box.

If you are using Qt5 (which doesn't have `QPromise`), you could write your own thread-save Awaitable, which would resume the awaiting coroutine from the main thread.

**Reply by SeeRich** [reply]
Thank you! I am currently using Qt6 so I will look into `QFuture/QPromise`.

**Reply by SeeRich** [reply]
@danvratil, sorry for the delay. I tried some of your advice but I have run into a brick wall (i.e. Unknown segmentation fault). Essentially, I converted my existing library code to use `boost::future` v. `std::future` simply because the boost version supports then continuations. Next I attempted to write a `QFutureAdapter` that would convert a `boost::future` into a `QFuture` which could then integrate and be used in my main application with QCoro. I feel like I am very close; however, as mentioned I am encountering an error and was hoping that you could help.

Below are three files that represent the minimal representation of my code base. This code is barrowed from QCoro networking example with the actual networking replaced by my boost + QFuture adapter attempt.

The system I am using is Ubuntu 22.04, GDB, GCC 11.2, boost version 1.80, Qt 6.3.2, QCoro 0.6.0. There are some additional notes listed in the `build.bash` script.

Any help would be greatly appreciated!

#### **`build.bash`**
```shell
# Run the program: `install\bin\qcoro-sandbox.exe`
# NOTES:
    # 1. You can get a copy of boost on Ubuntu 22.04 using `apt-get install libboost-all-dev`
    # 2. Don't forget to edit the CMAKE_PREFIX_PATH in the CMakeLists.txt to point to the boost,
    #   Qt and QCoro installation if installed in a custom location. If installed system wide, you
    #   should be able to simple remove these lines.
    # 3. This was tested on Ubuntu 22.04 using gdb and gcc 11.2
rm -rf build install
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=install -G Ninja
cmake --build build --config Debug -j12 --clean-first -v
cmake --install build --config Debug
```

#### **`CMakeLists.txt`**
```cmake
cmake_minimum_required(VERSION 3.18)

project(qcoro-sandbox LANGUAGES CXX)

# Add path to third-party libs (e.g. boost, qcoro here).
set(DEPS_PREFIX ${CMAKE_CURRENT_LIST_DIR}/../../../third-party/install)
set(CMAKE_PREFIX_PATH ${DEPS_PREFIX}/qcoro ${DEPS_PREFIX}/boost-release)
message(STATUS "CMAKE_PREFIX_PATH: ${CMAKE_PREFIX_PATH}")

# Search for required
find_package(Qt6 REQUIRED COMPONENTS Core Gui Widgets)
find_package(QCoro6 REQUIRED COMPONENTS Core Network)
find_package(Boost REQUIRED COMPONENTS headers program_options filesystem thread)

set(CMAKE_AUTOMOC ON)

add_executable(qcoro-sandbox Main.cpp)

target_include_directories(qcoro-sandbox PRIVATE ${CMAKE_CURRENT_LIST_DIR})

target_compile_features(qcoro-sandbox PRIVATE cxx_std_20)
target_compile_definitions(qcoro-sandbox PRIVATE BOOST_THREAD_VERSION=4 BOOST_THREAD_PROVIDES_FUTURE_CONTINUATION)
target_compile_options(qcoro-sandbox PRIVATE -Wall -Wextra -pedantic -Werror -Wsuggest-override)

target_link_libraries(qcoro-sandbox PRIVATE Qt6::Core Qt6::Gui Qt6::Widgets)
target_link_libraries(qcoro-sandbox PRIVATE QCoro::Core QCoro::Network)
target_link_libraries(qcoro-sandbox PRIVATE Boost::headers Boost::thread)

install(TARGETS qcoro-sandbox
    BUNDLE DESTINATION .)
```

#### **`Main.cpp`**
```cpp
#include "qcoro/qcoronetworkreply.h"
#include "qcoro/qcorotask.h"

#include <QApplication>
#ifdef __clang__
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored \
        "-Wdeprecated-enum-enum-conversion"  // QSizePolicy seems problematic...
#endif                                       // __clang__
#include <QBoxLayout>
#ifdef __clang__
    #pragma clang diagnostic pop
#endif  // __clang__
#include <QCoroFuture>
#include <QMainWindow>
#include <QMessageBox>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QProgressBar>
#include <QPushButton>

#include <boost/asio.hpp>
#include <boost/thread/future.hpp>

#include <chrono>
#include <iostream>
#include <memory>
#include <optional>
#include <string>

using namespace std::chrono_literals;
namespace net = boost::asio;  // from <boost/asio.hpp>

// Programming langauges
constexpr char raw_url[] = "http://www.wikidata.org/wiki/Special:EntityData/Q9143.json";
static const QUrl wikiUrl = QUrl{raw_url};

class AsyncManager
{
public:
    AsyncManager()
        : _work(net::require(_ioc.get_executor(), net::execution::outstanding_work.tracked))
    {
        _thread = std::thread([this]() { _ioc.run(); });
    }

    ~AsyncManager()
    {
        // Gracefully close down.
        // Reset the work object, this should block until all work is complete.
        _work = net::any_io_executor();
        if(!_ioc.stopped())
            _ioc.stop();
        _thread.join();
    }

    net::awaitable<std::string> async_op_coro(net::io_context& ioc,
                                              std::chrono::milliseconds timeRequired)
    {
        // Just simulate an asynchronous operation
        net::steady_timer timer(ioc);
        timer.expires_after(std::chrono::milliseconds(timeRequired));
        co_await timer.async_wait(net::use_awaitable);
        co_return "Async Operation Complete!!!!";
    }

    boost::future<std::string> asyncOperation(std::chrono::milliseconds timeRequired)
    {
        auto promise = std::make_shared<boost::promise<std::string>>();
        auto result = promise->get_future();
        // Spawn coroutine that will handle request.
        net::co_spawn(
            _ioc,
            async_op_coro(_ioc, timeRequired),
            [p = std::move(promise)](std::exception_ptr, auto&& result) { p->set_value(result); });
        return result;
    }

private:
    /// Thread responsible for servicing the requests
    std::thread _thread;
    /// I/O context responsible for handling all I/O.
    net::io_context _ioc;
    /// Executor that will prevent the service from stopping when it
    /// runs out of work.
    net::any_io_executor _work;
};

template <typename T>
class QFutureAdapter
{
public:
    QFutureAdapter(boost::future<T>&& fut) : _fut(std::move(fut))
    {
        std::cout << "QFutureAdapter CTOR" << std::endl;
    }

    ~QFutureAdapter() { std::cout << "QFutureAdapter DTOR" << std::endl; }

    auto operator()() -> QFuture<T>
    {
        auto promise = std::make_shared<QPromise<T>>();
        auto future = promise->future();
        _fut.then([p = promise](boost::future<T> completeFut) -> void {
            std::cout << "then callback..." << std::endl;
            try {
                p->addResult(completeFut.get());
            } catch(const std::exception& e) {
                std::cout << "then callback error: " << e.what() << std::endl;
            }
            std::cout << "then callback exit..." << std::endl;
        });
        return future;
    }

private:
    boost::future<T> _fut;
};

class MainWindow : public QMainWindow
{
    Q_OBJECT
public:
    MainWindow()
    {
        mPb = new QProgressBar();
        mPb->setVisible(false);
        mPb->setMinimumWidth(200);
        mPb->setMinimum(0);
        mPb->setMaximum(0);

        mBtn = new QPushButton(tr("Start Download"));
        connect(mBtn, &QPushButton::clicked, this, &MainWindow::start);

        auto* vbox = new QVBoxLayout();
        vbox->addStretch(1);
        vbox->addWidget(mPb);
        vbox->addWidget(mBtn);
        vbox->addStretch(1);

        auto* hbox = new QHBoxLayout();
        hbox->addStretch(1);
        hbox->addLayout(vbox);
        hbox->addStretch(1);

        QWidget* w = new QWidget;
        w->setLayout(hbox);
        setCentralWidget(w);

        _async_manager = std::make_shared<AsyncManager>();
    }

private Q_SLOTS:
    QCoro::Task<void> start()
    {
        try {
            mPb->setVisible(true);
            mBtn->setEnabled(false);
            mBtn->setText(tr("Downloading ..."));

            auto adapter = QFutureAdapter<std::string>(_async_manager->asyncOperation(3000ms));
            auto result = co_await adapter();

            std::cout << "ASYNC RESULT: " << result << std::endl;

            mPb->setVisible(false);
            mBtn->setEnabled(true);
            mBtn->setText(tr("Done, download again"));

            std::cout << "Done Downloading..." << std::endl;
            co_return;
        } catch(const std::exception& e) {
            std::cout << "Start error: " << e.what() << std::endl;
        }
    }

private:
    std::shared_ptr<AsyncManager> _async_manager;
    QPushButton* mBtn = {};
    QProgressBar* mPb = {};
};

int main(int argc, char** argv)
{
    QApplication app(argc, argv);
    MainWindow window;
    window.showNormal();
    return app.exec();
}

#include "Main.moc"
```

**Reply by danvratil** [reply]
Hi,

I think you haven't done anything wrong - from running the code under ASAN, I suspect you found a bug in QCoro :-) I'll get back to you as I find out more.

**Reply by SeeRich** [reply]
Ok, thank you very much! Let me know if I can be of any assistance.

**Reply by danvratil** [reply]
So, upon further investigation, the main bug is in your code, there's a bug in my code too and also there's a weird thing in Qt :-)

tl;dr: you are missing `p->finish()` in the `_fut.then()` callback i n`QFutureAdapter::operator()` after adding the result to the promise, so that the `await`ed future completes and `start()` resumes as expected.

Full explanation:

In the `_fut.then()` callback inside `QFutureAdapter::operator()` you call `p->addResult()`, which "sends" the result to the `QFuture` that you are `co_await`ing inside `start()`. However, that does not resume the `start()` coroutine, because just adding a result into promise/future doesn't make the future finished (you can add more results). So `start()` is stil suspeneded at this point.

What happens next is that the callback in `_fut.then()` reaches the end and is destroyed. This destroys the last reference to the `QPromise` shared pointer and thus it destroys the `QPromise`, which [according to documentation](https://doc.qt.io/qt-6/qpromise.html#dtor.QPromise), makes the associated future `cancelled`, if it currently is in running state. The future cancellation actually triggers handling inside `QCoroFuture` and causes the `start()` coroutine to be resumed. You even get to extract the result from it (which is OK), but if you were to check the state of the future returned from `adapter()` call, you would see that the future was cancelled, rather than finished.

Now for a reason that I do not understand (and what seems to be a bug or at least undocumented behavior), when QPromise is destroyed while not in finished state, it will send out the cancellation notification, followed by finish notification. The finish notification triggers the same code in QCoro again (this is the bug in QCoro), which then tries to resume the `start()` coroutine for the second time and crashes, because the `start()` coroutine has finished in the meantime, and `QCoro` is by now holding dangling reference the coroutine.

If you fix your code to mark the promise as finished, it will work as expected. I'll go and fix QCoro to not misbehave like this. And I should investigate the Qt behavior in more depth.

**Reply by danvratil** [reply]
Clarification: after reading the entire `QFuture` documentation page, while not explicitly stated anywhere, the future can be both canceled and finished, as implied by the docs for `QFuture::waitForFinished()`:
> Waits for the asynchronous computation to finish (including [cancel](https://doc.qt.io/qt-6/qfuture.html#cancel)()ed computations), i.e. until [isFinished](https://doc.qt.io/qt-6/qfuture.html#isFinished)() returns true.

**Reply by danvratil** [reply]
Thinking about the comment above, the main bug really is in QCoro which reacts to both `canceled()` and `finished()` signals, so in case of cancellation, QCoro triggers twice, while it should only trigger once after the `finished()` signal, to behave consistently with `QFuture<T>::waitForFinished()`.

Anyway, your code should probably be fixed too ;-)

**Reply by danvratil** [reply]
And to make the whole thing even more messy is the fact that QPromise cancelling on destruction has been introduced in Qt 6.3.1 and backported to 6.2.5, so your code would hang indefinitely when running against older Qt 6.

**Reply by SeeRich** [reply]
Wow, that's what I get for not reading all of the `QFuture`/`QPromise` documentation carefully. Thank you so much for your detailed investigation! I made the `p->finish()` change you suggested and everything is working now. Now I will work on handling the case where the `boost::asio` coroutine throws and exception or experiences a timeout. Thank you again for your time!

**Reply by SeeRich** [reply]
Thankfully, the Qt part of my application is pretty new so I am free to keep moving forward with the new Qt 6 releases.

**Reply by SeeRich** [reply]
Currently, are the exceptions from a future i.e. (`QPromise::setException()`) propagated to the calling code? I made some simple changes to my code base to test this out but I can't seem to catch the exception in my `QCoro::Task`. The `QFuture`/`QPromise` documentation seems to indicate this is done with the continuation `onFailed`; however, it doesn't seem that qcoro is using this?

Here are the changes I made:
```c++
auto operator()() -> QFuture<T>
{
    auto promise = std::make_shared<QPromise<T>>();
    auto future = promise->future();
    _fut.then([p = promise](boost::future<T> completeFut) -> void {
        std::cout << "then callback..." << std::endl;
        try {
            throw std::runtime_error("TESTING EXCEPTION");
            p->addResult(std::move(completeFut.get()));
            p->finish();
        } catch(const std::exception& e) {
            std::cout << "then callback error: " << e.what() << std::endl;
            p->setException(std::current_exception());
            // p->finish();
        }
        std::cout << "then callback exit..." << std::endl;
    });
    return future;
}
```

and

```c++
try {
    auto adapter = QFutureAdapter<std::string>(_async_manager->asyncOperation(300ms));
    auto result = co_await adapter();
    std::cout << "ASYNC RESULT: " << result << std::endl;
} catch(const std::exception& e) {
    std::cout << "HANDLING EXCEPTION: " << e.what() << std::endl;
    co_return;
}
```

**Reply by danvratil** [reply]
QCoro supports only exceptions normally thrown inside a coroutine. Explicit support for rethrowing an exception stored in QFuture sounds like a sensible feature to have, and should be fairly straightforward to implement inside QCoroFuture. I'll look into it in the next days.

**Reply by danvratil** [reply]
You need to call `p->finish()` in the exception-handling path, otherwise the Future will never finish (and re-throw the exception).

On QCoro side, there's a simple change needed, I already merged it into `main` and added a bunch of tests to make sure it keeps working :)

**Reply by SeeRich** [reply]
This works exactly as expected. Thank you for all the support!

Below is a copy of the final working code with exception handling in case it is helpful to someone else.

```c++
#include "qcoro/qcoronetworkreply.h"
#include "qcoro/qcorotask.h"

#include <QApplication>
#ifdef __clang__
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored \
        "-Wdeprecated-enum-enum-conversion"  // QSizePolicy seems problematic...
#endif                                       // __clang__
#include <QBoxLayout>
#ifdef __clang__
    #pragma clang diagnostic pop
#endif  // __clang__
#include <QCoroFuture>
#include <QMainWindow>
#include <QMessageBox>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QProgressBar>
#include <QPushButton>

#include <boost/asio.hpp>
#include <boost/thread/future.hpp>

#include <chrono>
#include <iostream>
#include <memory>
#include <optional>
#include <string>

using namespace std::chrono_literals;
namespace net = boost::asio;  // from <boost/asio.hpp>

// Programming langauges
constexpr char raw_url[] = "http://www.wikidata.org/wiki/Special:EntityData/Q9143.json";
static const QUrl wikiUrl = QUrl{raw_url};

/// Convenience function to convert a std::exception_ptr to a boost::exception_ptr
boost::exception_ptr convert_exception_s2b(std::exception_ptr ex)
{
    try {
        std::rethrow_exception(ex);
    } catch(...) {
        return boost::current_exception();
    }
}

/// Convenience function to convert a boost::exception_ptr to a std::exception_ptr
std::exception_ptr convert_exception_b2s(boost::exception_ptr ex)
{
    try {
        boost::rethrow_exception(ex);
    } catch(...) {
        return std::current_exception();
    }
}

class AsyncManager
{
public:
    AsyncManager()
        : _work(net::require(_ioc.get_executor(), net::execution::outstanding_work.tracked))
    {
        _thread = std::thread([this]() { _ioc.run(); });
    }

    ~AsyncManager()
    {
        // Gracefully close down.
        // Reset the work object, this should block until all work is complete.
        _work = net::any_io_executor();
        if(!_ioc.stopped())
            _ioc.stop();
        _thread.join();
    }

    net::awaitable<std::string> async_op_coro(net::io_context& ioc,
                                              std::chrono::milliseconds timeRequired,
                                              bool simulateException)
    {
        // Just simulate an asynchronous operation
        net::steady_timer timer(ioc);
        timer.expires_after(std::chrono::milliseconds(timeRequired));
        co_await timer.async_wait(net::use_awaitable);
        if(simulateException)
            BOOST_THROW_EXCEPTION(std::runtime_error("Async Error Timeout"));
        co_return "Async Operation Complete!!!!";
    }

    boost::future<std::string> asyncOperation(std::chrono::milliseconds timeRequired,
                                              bool simulateException)
    {
        auto promise = std::make_shared<boost::promise<std::string>>();
        auto result = promise->get_future();
        // Spawn coroutine that will handle request.
        net::co_spawn(_ioc,
                      async_op_coro(_ioc, timeRequired, simulateException),
                      [p = std::move(promise)](std::exception_ptr e, auto&& result) {
                          if(!e) {
                              p->set_value(result);
                          } else {
                              // Convert exception types to maintain exception information
                              auto bexp = convert_exception_s2b(e);
                              p->set_exception(bexp);
                          }
                      });
        return result;
    }

private:
    /// Thread responsible for servicing the requests
    std::thread _thread;
    /// I/O context responsible for handling all I/O.
    net::io_context _ioc;
    /// Executor that will prevent the service from stopping when it
    /// runs out of work.
    net::any_io_executor _work;
};

template <typename T>
class QFutureAdapter
{
public:
    QFutureAdapter(boost::future<T>&& fut) : _fut(std::move(fut)) {}

    ~QFutureAdapter() {}

    auto operator()() -> QFuture<T>
    {
        auto promise = std::make_shared<QPromise<T>>();
        auto future = promise->future();
        _fut.then([p = promise](boost::future<T> completeFut) -> void {
            std::cout << "then callback..." << std::endl;
            if(auto exp = completeFut.get_exception_ptr()) {
                auto sexp = convert_exception_b2s(exp);
                p->setException(sexp);
            } else {
                p->addResult(std::move(completeFut.get()));
            }
            // Always mark as finished.
            p->finish();
            std::cout << "then callback exit..." << std::endl;
        });
        return future;
    }

private:
    boost::future<T> _fut;
};

class MainWindow : public QMainWindow
{
    Q_OBJECT
public:
    MainWindow()
    {
        mPb = new QProgressBar();
        mPb->setVisible(false);
        mPb->setMinimumWidth(200);
        mPb->setMinimum(0);
        mPb->setMaximum(0);

        mBtn = new QPushButton(tr("Start Download"));
        connect(mBtn, &QPushButton::clicked, this, &MainWindow::start);

        auto* vbox = new QVBoxLayout();
        vbox->addStretch(1);
        vbox->addWidget(mPb);
        vbox->addWidget(mBtn);
        vbox->addStretch(1);

        auto* hbox = new QHBoxLayout();
        hbox->addStretch(1);
        hbox->addLayout(vbox);
        hbox->addStretch(1);

        QWidget* w = new QWidget;
        w->setLayout(hbox);
        setCentralWidget(w);

        _async_manager = std::make_shared<AsyncManager>();
    }

private Q_SLOTS:
    QCoro::Task<void> start()
    {
        try {
            mPb->setVisible(true);
            mBtn->setEnabled(false);
            mBtn->setText(tr("Downloading ..."));

            // Make request expecting success
            try {
                auto adapter =
                    QFutureAdapter<std::string>(_async_manager->asyncOperation(300ms, false));
                auto result = co_await adapter();
                std::cout << "ASYNC RESULT: " << result << std::endl;
            } catch(const boost::exception& e) {
                std::cout << boost::diagnostic_information(e) << std::endl;
            } catch(const std::exception& se) {
                std::cout << se.what() << std::endl;
            }

            // Make request expecting error
            try {
                auto adapter =
                    QFutureAdapter<std::string>(_async_manager->asyncOperation(300ms, true));
                auto result = co_await adapter();
                std::cout << "ASYNC RESULT: " << result << std::endl;
            } catch(const boost::exception& e) {
                std::cout << boost::diagnostic_information(e) << std::endl;
            } catch(const std::exception& se) {
                std::cout << se.what() << std::endl;
            }

            mPb->setVisible(false);
            mBtn->setEnabled(true);
            mBtn->setText(tr("Done, download again"));

            co_return;
        } catch(const std::exception& e) {
            std::cout << "Start error: " << e.what() << std::endl;
        }
    }

private:
    std::shared_ptr<AsyncManager> _async_manager;
    QPushButton* mBtn = {};
    QProgressBar* mPb = {};
};

int main(int argc, char** argv)
{
    QApplication app(argc, argv);
    MainWindow window;
    window.showNormal();
    return app.exec();
}

#include "Main.moc"
```

**Comment by SeeRich** [comment]
Hey, I just want to say that your help in this discussion has helped me refactor my application tremendously, thank you!

I have one additional question related to this setup. The application that I am integrating QCoro requires that exceptions occurring in asynchronous contexts must propagate to the main thread and cause the application to close so that the stack trace and exception can be logged and reported.

All that to say I am experiencing somewhat strange behavior when using coroutines that throw exceptions.

The following code is a refactored version of the above to demonstrate the issue. I expect that calling both error functions (i.e. `coAwaitErrorFunction` & `nonCoAwaitErrorFunction`) should be caught by the try/catch block around the main application. However, the function that does something useful and has a `co_await` keyword (i.e. Function 1), seems to just ignore the exception.

```c++
#include "qcoro/qcoronetworkreply.h"
#include "qcoro/qcorotask.h"

#include <QApplication>
#ifdef __clang__
    #pragma clang diagnostic push
    #pragma clang diagnostic ignored \
        "-Wdeprecated-enum-enum-conversion"  // QSizePolicy seems problematic...
#endif                                       // __clang__
#include <QBoxLayout>
#ifdef __clang__
    #pragma clang diagnostic pop
#endif  // __clang__
#include <QCoroFuture>
#include <QMainWindow>
#include <QMessageBox>
#include <QNetworkAccessManager>
#include <QNetworkReply>
#include <QNetworkRequest>
#include <QProgressBar>
#include <QPushButton>

#include <boost/asio.hpp>
#include <boost/thread/future.hpp>

#include <chrono>
#include <iostream>
#include <memory>
#include <optional>
#include <string>

using namespace std::chrono_literals;
namespace net = boost::asio;  // from <boost/asio.hpp>

/// Convenience function to convert a std::exception_ptr to a boost::exception_ptr
boost::exception_ptr convert_exception_s2b(std::exception_ptr ex)
{
    try {
        std::rethrow_exception(ex);
    } catch(...) {
        return boost::current_exception();
    }
}

/// Convenience function to convert a boost::exception_ptr to a std::exception_ptr
std::exception_ptr convert_exception_b2s(boost::exception_ptr ex)
{
    try {
        boost::rethrow_exception(ex);
    } catch(...) {
        return std::current_exception();
    }
}

class AsyncManager
{
public:
    AsyncManager()
        : _work(net::require(_ioc.get_executor(), net::execution::outstanding_work.tracked))
    {
        _thread = std::thread([this]() { _ioc.run(); });
    }

    ~AsyncManager()
    {
        // Gracefully close down.
        // Reset the work object, this should block until all work is complete.
        _work = net::any_io_executor();
        if(!_ioc.stopped())
            _ioc.stop();
        _thread.join();
    }

    net::awaitable<std::string> async_op_coro(net::io_context& ioc,
                                              std::chrono::milliseconds timeRequired,
                                              bool simulateException)
    {
        // Just simulate an asynchronous operation
        net::steady_timer timer(ioc);
        timer.expires_after(std::chrono::milliseconds(timeRequired));
        co_await timer.async_wait(net::use_awaitable);
        if(simulateException)
            BOOST_THROW_EXCEPTION(std::runtime_error("Async Error Timeout"));
        co_return "Async Operation Complete!!!!";
    }

    boost::future<std::string> asyncOperation(std::chrono::milliseconds timeRequired,
                                              bool simulateException)
    {
        auto promise = std::make_shared<boost::promise<std::string>>();
        auto result = promise->get_future();
        // Spawn coroutine that will handle request.
        net::co_spawn(_ioc,
                      async_op_coro(_ioc, timeRequired, simulateException),
                      [p = std::move(promise)](std::exception_ptr e, auto&& result) {
                          if(!e) {
                              p->set_value(result);
                          } else {
                              // Convert exception types to maintain exception information
                              auto bexp = convert_exception_s2b(e);
                              p->set_exception(bexp);
                          }
                      });
        return result;
    }

private:
    /// Thread responsible for servicing the requests
    std::thread _thread;
    /// I/O context responsible for handling all I/O.
    net::io_context _ioc;
    /// Executor that will prevent the service from stopping when it
    /// runs out of work.
    net::any_io_executor _work;
};

template <typename T>
class QFutureAdapter
{
public:
    QFutureAdapter(boost::future<T>&& fut) : _fut(std::move(fut)) {}

    ~QFutureAdapter() {}

    auto operator()() -> QFuture<T>
    {
        auto promise = std::make_shared<QPromise<T>>();
        auto future = promise->future();
        _fut.then([p = promise](boost::future<T> completeFut) -> void {
            std::cout << "then callback..." << std::endl;
            if(auto exp = completeFut.get_exception_ptr()) {
                auto sexp = convert_exception_b2s(exp);
                p->setException(sexp);
            } else {
                p->addResult(std::move(completeFut.get()));
            }
            // Always mark as finished.
            p->finish();
            std::cout << "then callback exit..." << std::endl;
        });
        return future;
    }

private:
    boost::future<T> _fut;
};

class MainWindow : public QMainWindow
{
    Q_OBJECT
public:
    MainWindow()
    {
        mPb = new QProgressBar();
        mPb->setVisible(false);
        mPb->setMinimumWidth(200);
        mPb->setMinimum(0);
        mPb->setMaximum(0);

        mBtn = new QPushButton(tr("Call Error Function 1"));
        connect(mBtn, &QPushButton::clicked, this, &MainWindow::coAwaitErrorFunction);

        mBtn2 = new QPushButton(tr("Call Error Function 2"));
        connect(mBtn2, &QPushButton::clicked, this, &MainWindow::nonCoAwaitErrorFunction);

        auto* vbox = new QVBoxLayout();
        vbox->addStretch(1);
        vbox->addWidget(mPb);
        vbox->addWidget(mBtn);
        vbox->addWidget(mBtn2);
        vbox->addStretch(1);

        auto* hbox = new QHBoxLayout();
        hbox->addStretch(1);
        hbox->addLayout(vbox);
        hbox->addStretch(1);

        QWidget* w = new QWidget;
        w->setLayout(hbox);
        setCentralWidget(w);

        _async_manager = std::make_shared<AsyncManager>();
    }

private Q_SLOTS:
    QCoro::Task<void> coAwaitErrorFunction()
    {
        try {
            mPb->setVisible(true);
            mBtn->setEnabled(false);
            mBtn2->setEnabled(false);
            mBtn->setText(tr("Downloading ..."));

            // Make request expecting error
            auto adapter = QFutureAdapter<std::string>(_async_manager->asyncOperation(300ms, true));
            auto result = co_await adapter();
            std::cout << "ASYNC RESULT: " << result << std::endl;
            co_return;
        } catch(const std::exception& e) {
            std::cout << "Start error: " << e.what() << std::endl;
            mPb->setVisible(false);
            mBtn->setEnabled(true);
            mBtn2->setEnabled(true);
            mBtn->setText(tr("Done, download again"));
            // Re throw the exception
            throw;
        }
    }

    QCoro::Task<void> nonCoAwaitErrorFunction() { throw std::runtime_error("I'm a bad coroutine"); }

private:
    std::shared_ptr<AsyncManager> _async_manager;
    QPushButton* mBtn = {};
    QPushButton* mBtn2 = {};
    QProgressBar* mPb = {};
};

int main(int argc, char** argv)
{
    QApplication app(argc, argv);
    try {
        MainWindow window;
        window.showNormal();
        return app.exec();
    } catch(...) {
        std::cout << "CHANCE TO HANDLE GLOBAL UNEXPECTED ERRORS" << std::endl;
    }
}

#include "Main.moc"
```

**Reply by danvratil** [reply]
I'm glad that I can help and that QCoro works for you nicely 🙂

I did not try to run your code, and won't be able to investigate in depth until next week, but from your description I think you are hitting a known limitation - non-awaited coroutines don't rethrow exceptions (#97)

Basically, when a coroutine throws, the exception gets stored, the coroutine is finished immediately, and if someone is awaiting the coroutine, they get resumed. The resumed awaiter tries to obtain the result of the coroutine and at this point, the stored exception gets rethrown inside the awaiter.

In your case, the exception is thrown inside the `coAwaitErrorFunction` which is a coroutine slot invoked from the event loop. However the event loop doesn't explicitly co_await the slot, it just discards the returned `QCoro::Task<>` right away. Thus, the exception is stored, but is never rethrown, because there's nothing trying to read the result of the coroutine slot.

There's more in issue #97, and there's also an experimental branch that makes the behavior of throwing exception from non-awaited coroutine optional. However I'm quite unhappy with the implementation and haven't yet figured out another way.
